flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSink.java

```diff
@@ -0,0 +1,159 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.batch.connectors.hive;
+
+import org.apache.flink.api.common.io.OutputFormat;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.table.catalog.exceptions.CatalogException;
+import org.apache.flink.table.catalog.hive.client.HiveMetastoreClientFactory;
+import org.apache.flink.table.catalog.hive.client.HiveMetastoreClientWrapper;
+import org.apache.flink.table.catalog.hive.client.HiveShimLoader;
+import org.apache.flink.table.catalog.hive.descriptors.HiveCatalogValidator;
+import org.apache.flink.table.sinks.OutputFormatTableSink;
+import org.apache.flink.table.sinks.TableSink;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.guava18.com.google.common.base.Joiner;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.MetaStoreUtils;
+import org.apache.hadoop.hive.metastore.Warehouse;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.thrift.TException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+/**
+ * Table sink to write to Hive tables.
+ */
+public class HiveTableSink extends OutputFormatTableSink<Row> {
+
+	private final JobConf jobConf;
+	private final RowTypeInfo rowTypeInfo;
+	private final String dbName;
+	private final String tableName;
+	private final List<String> partitionColumns;
+	private final String hiveVersion;
+
+	// TODO: need OverwritableTableSink to configure this
+	private boolean overwrite = false;
+
+	public HiveTableSink(JobConf jobConf, RowTypeInfo rowTypeInfo, String dbName, String tableName,
+			List<String> partitionColumns) {
+		this.jobConf = jobConf;
+		this.rowTypeInfo = rowTypeInfo;
+		this.dbName = dbName;
+		this.tableName = tableName;
+		this.partitionColumns = partitionColumns;
+		hiveVersion = jobConf.get(HiveCatalogValidator.CATALOG_HIVE_VERSION, HiveShimLoader.getHiveVersion());
+	}
+
+	@Override
+	public OutputFormat<Row> getOutputFormat() {
+		boolean isPartitioned = partitionColumns != null && !partitionColumns.isEmpty();
+		// TODO: need PartitionableTableSink to decide
+		boolean isDynamicPartition = isPartitioned;
```

Comment:
this means the table will be dynamically partitioned as long as it's partitioned table. Is this right? If that's not true, then we should set it to false until it's supported, right?

Reply:
That's true. Currently we don't have the interface to set static partition spec. Therefore we only support either non-partitioned or fully dynamic partition insert.